src/blueprint.zig

```diff
@@ -0,0 +1,254 @@
+const std = @import("std");
+
+const main = @import("main.zig");
+const Compression = main.utils.Compression;
+const ZonElement = @import("zon.zig").ZonElement;
+const vec = main.vec;
+const Vec3i = vec.Vec3i;
+
+const Array3D = main.utils.Array3D;
+const Block = main.blocks.Block;
+const NeverFailingAllocator = main.heap.NeverFailingAllocator;
+const User = main.server.User;
+
+pub const GameIdToBlueprintIdMapType = std.AutoHashMap(u16, u16);
+const BlockIdSizeType = u32;
+const BlockStorageType = u32;
+
+const BinaryWriter = main.utils.BinaryWriter;
+const BinaryReader = main.utils.BinaryReader;
+
+pub const blueprintVersion = 0;
+
+pub const BlueprintCompression = enum(u16) {
+	deflate,
+};
+
+pub const Blueprint = struct {
+	blocks: Array3D(Block),
+
+	pub fn init(allocator: NeverFailingAllocator) Blueprint {
+		return .{.blocks = .init(allocator, 0, 0, 0)};
+	}
+	pub fn deinit(self: *Blueprint, allocator: NeverFailingAllocator) void {
+		self.blocks.deinit(allocator);
+	}
+	pub fn clone(self: *Blueprint, allocator: NeverFailingAllocator) Blueprint {
+		var new = Blueprint.init(allocator);
+		new.blocks = self.blocks.clone(allocator);
+		return new;
+	}
+	const CaptureResult = union(enum) {
+		success: Blueprint,
+		failure: struct {x: i32, y: i32, z: i32, message: []const u8},
+	};
+
+	pub fn capture(allocator: NeverFailingAllocator, pos1: Vec3i, pos2: Vec3i) CaptureResult {
+		var self = Blueprint{.blocks = undefined};
+
+		const startX = @min(pos1[0], pos2[0]);
+		const startY = @min(pos1[1], pos2[1]);
+		const startZ = @min(pos1[2], pos2[2]);
+
+		const endX = @max(pos1[0], pos2[0]);
+		const endY = @max(pos1[1], pos2[1]);
+		const endZ = @max(pos1[2], pos2[2]);
+
+		const sizeX: u32 = @intCast(endX - startX + 1);
+		const sizeY: u32 = @intCast(endY - startY + 1);
+		const sizeZ: u32 = @intCast(endZ - startZ + 1);
+
+		self.blocks = .init(allocator, sizeX, sizeY, sizeZ);
+
+		for(0..sizeX) |x| {
+			const worldX = startX +% @as(i32, @intCast(x));
+
+			for(0..sizeY) |y| {
+				const worldY = startY +% @as(i32, @intCast(y));
+
+				for(0..sizeZ) |z| {
+					const worldZ = startZ +% @as(i32, @intCast(z));
+
+					const maybeBlock = main.server.world.?.getBlock(worldX, worldY, worldZ);
```

Comment:
Ideally you would get the chunk, and then add all blocks from it in one go (which would require 6 loops).
However I can understand that you may not want to go through the trouble, and at the moment that is far from becoming a bottleneck (especially for pasting)